#!/bin/bash
#SBATCH --time=12:00:00 --mem=10GB

# This scripts assumes you have run the ipsc_preproccess_pipeline first (https://github.com/BennyStrobes/ipsc_preprocess_pipeline)
# And have save the results here:
preprocess_dir="/project2/gilad/bstrober/ipsc_differentiation/preprocess/"


##########################################################################
# Input Files
##########################################################################

# Heterozygous probability genotype file for all cell_lines in our analysis
# These heterozygous probabilities come from impute2
het_prob_genotype_file=$preprocess_dir"genotype/YRI_het_prob_genotype.vcf"

# Dosage genotype for all cell lines in our analysis
dosage_genotype_file=$preprocess_dir"genotype/YRI_genotype.vcf"

# Processed Allelic counts
# Assuming heterozygous threshold of .999 in terms of calling heterozygous sites
allelic_counts_file=$preprocess_dir"processed_allelic_counts/allelic_counts_gene_mapped_het_prob_999.txt"

# SVA loadings for each of the samples
# This is our covariate file
sva_loading_file=$preprocess_dir"covariates/sva_loadings.txt"


# PCA loadings for our samples (done independently in each time step)
#  Therefor, there is one file for each time step
pca_loading_file_stem=$preprocess_dir"covariates/pca_loadings_time_step_independent_"

# Processed total expression data
# This is the quantile normalized expression data
quantile_normalized_expression=$preprocess_dir"processed_total_expression/quantile_normalized.txt"

#  Processed total expression data that was processed independently at each time point
# This is the quantile normalized data
quantile_normalized_time_step_independent_expression=$preprocess_dir"processed_total_expression/time_step_independent_quantile_normalized.txt"

# Processed and corrected total expression data
# This the quantile normalized expression data with the SVA latent factors regressed out
corrected_quantile_normalized_expression=$preprocess_dir"processed_total_expression/quant_expr_sva_corrected.txt"


# Gencode hg19 gene annotation file
gencode_gene_annotation_file="/project2/gilad/bstrober/ipsc_differentiation/preprocess_input_data/gencode.v19.annotation.gtf.gz"

# QTLs generated when all ipscs were used (generated by nick banovich)
full_ipsc_qtl_data="/project2/gilad/bstrober/ipsc_differentiation/preprocess_input_data/permutations.all.RNAseq_run.fixed.txt.gz.bh.txt"

# eQTLS generated in GTEx V7 for heart left ventricle
# downloaded from https://www.gtexportal.org/home/datasets/ on 10/30/17
full_heart_eqtl_data="/project2/gilad/bstrober/ipsc_differentiation/preprocess_input_data/Heart_Left_Ventricle.v7.egenes.txt"


# Downloaded from http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz on 10/20/17
# Required by WASP
chrom_info_file="/project2/gilad/bstrober/ipsc_differentiation/preprocess_input_data/chromInfo.txt"

##########################################################################
# Output Directories (Assumes these directories exist before script starts)
##########################################################################
# Root directory for ipsc qtl pipelines
ipsc_qtl_root="/project2/gilad/bstrober/ipsc_differentiation/qtl_pipelines/"

# Output directory for running ase qtl analysis at each time step independently
independent_time_step_ase_qtl_dir=$ipsc_qtl_root"independent_time_step_ase_qtl/"

# Output directory for visualizations/plots related to running ase qtl analysis at each time step independently
visualize_independent_time_step_ase_qtl_dir=$ipsc_qtl_root"visualize_independent_time_step_ase_qtl/"


# Output directory for running  eqtl analysis at each time step independently
independent_time_step_eqtl_dir=$ipsc_qtl_root"independent_time_step_eqtl/"

# Output directory for visualizations/plots related to running eqtl analysis at each time step independently
visualize_independent_time_step_eqtl_dir=$ipsc_qtl_root"visualize_independent_time_step_eqtl/"

# Output directory for optimizing number of PCs to regress out for eqtl analysis
optimize_number_pcs_dir=$ipsc_qtl_root"optimize_number_pcs/"

# Output directory for running dynamic qtl analysis on simulated data
simulated_dynamic_qtl_dir=$ipsc_qtl_root"simulated_dynamic_qtl/"







############################################################################
### Run independent time step ase-qtl analysis
############################################################################

##################
# Model Parameters
##################
# Cis eqtl distance (EAGLE used 200 KB)
distance="100000"
# Minimum number of reads on heterozygous site to consider a sample valid
min_reads="2"
# Minimum fraction of heterozygous samples that show biallelic expression
min_fraction_biallelic=".7"
# Minimum number of heterozygous samples for us to run the test
min_samples_per_time_step="5"
# Minimum fraction of valid samples with less popular version of regulatory variant (homozygous reg variant vs heterozygous reg variant)
min_fraction_in_test_group=".2"
# Statistical test to use in evaluating aseQTL
# Options available are "wilcoxon" and "linear_regression
statistical_test="wilcoxon"
# Normalization method (only applicable if statistical_test="linear_regression", otherwise will be ignored)
# Options available are "standardize" and "quantile_normalize"
normalization_method="na"
###############################

##################
# Run different versions of model
##################


## Run wilcoxon test
statistical_test="wilcoxon"
normalization_method="na"
output_file_stem="independent_time_step_ase_qtl__distance_"$distance"_min_reads_"$min_reads"_min_fraction_biallelic_"$min_fraction_biallelic"_min_samples_"$min_samples_per_time_step"_min_fraction_in_test_group_"$min_fraction_in_test_group"_"$statistical_test"_normalization_method_"$normalization_method"_"
if false; then
sh independent_time_step_ase_qtl_driver.sh $het_prob_genotype_file $allelic_counts_file $output_file_stem $min_reads $min_samples_per_time_step $min_fraction_in_test_group $distance $statistical_test $normalization_method $min_fraction_biallelic $visualize_independent_time_step_ase_qtl_dir $independent_time_step_ase_qtl_dir
fi




# Run linear regression test (standardizing the response variable)
statistical_test="linear_regression"
normalization_method="standardize"
output_file_prefix=$independent_time_step_ase_qtl_dir"independent_time_step_qtl_distance_"$distance"_min_reads_"$min_reads"_min_fraction_biallelic_"$min_fraction_biallelic"_min_samples_"$min_samples_per_time_step"_min_fraction_in_test_group_"$min_fraction_in_test_group"_"$statistical_test"_normalization_method_"$normalization_method"_"
if false; then
sbatch independent_time_step_ase_qtl_driver.sh $het_prob_genotype_file $allelic_counts_file $output_file_prefix $min_reads $min_samples_per_time_step $min_fraction_in_test_group $distance $statistical_test $normalization_method $min_fraction_biallelic $visualize_independent_time_step_ase_qtl_dir
fi

if false; then
# Run linear regression test (quantile normalizing the response variable)
normalization_method="quantile_normalize"
output_file_prefix=$independent_time_step_ase_qtl_dir"independent_time_step_qtl_distance_"$distance"_min_reads_"$min_reads"_min_fraction_biallelic_"$min_fraction_biallelic"_min_samples_"$min_samples_per_time_step"_min_fraction_in_test_group_"$min_fraction_in_test_group"_"$statistical_test"_normalization_method_"$normalization_method"_"
fi















##################################################################
### Run independent time step e-qtl analysis
##################################################################
############# Parameters #################
# Cis eqtl distance (EAGLE used 200 KB)
distance="50000"

# Minimum fraction of valid samples with less popular version of regulatory variant (homozygous reg variant vs heterozygous reg variant)
maf_cutoff=".1"

# The way in which we normalize the data
# Currently implemented for:
#####1. 'none'
#####2. 'standardize'
#####3. 'gaussian_projection'
normalization_method="none"

# The way in which the data was prepared
# Currently implemented for:
#####1. 'time_step_aggregrated'  (quantile normalization and hidden factor correction was done across all samples. But regressed at each time step seperately)
#####2. 'time_step_independent'  (quantile normalization and hidden factor correction was done independently at each time step)
#####3. 'time_step_aggregrated_all_regressed' (quantile normalization and hidden factor correction was done across all samples and regressed across all samples)
data_prep_version="time_step_independent"


### We first need to optimize the number of principal components to use
time_step_for_optimization_of_pcs="0"
if false; then
sh optimize_number_pcs_for_eqtl_driver.sh $dosage_genotype_file $quantile_normalized_expression $gencode_gene_annotation_file $optimize_number_pcs_dir $distance $maf_cutoff $time_step_for_optimization_of_pcs $normalization_method $data_prep_version $quantile_normalized_time_step_independent_expression $pca_loading_file_stem$time_step_for_optimization_of_pcs".txt" $sva_loading_file
fi

# The number of PCs to inlcude in the model 
num_pcs="1"

if false; then
for time_step in $(seq 0 15); do
    sbatch independent_time_step_eqtl_driver.sh $dosage_genotype_file $quantile_normalized_expression $gencode_gene_annotation_file $independent_time_step_eqtl_dir $distance $maf_cutoff $time_step $visualize_independent_time_step_eqtl_dir $full_ipsc_qtl_data $full_heart_eqtl_data $normalization_method $data_prep_version $num_pcs $quantile_normalized_time_step_independent_expression $pca_loading_file_stem$time_step".txt" $sva_loading_file
done
fi


if false; then
num_pcs="3"
Rscript visualize_eqtls_across_time_steps.R $visualize_independent_time_step_eqtl_dir $distance $maf_cutoff $normalization_method $independent_time_step_eqtl_dir $data_prep_version $num_pcs

data_prep_version="time_step_independent"


num_pcs="3"
Rscript visualize_eqtls_across_time_steps.R $visualize_independent_time_step_eqtl_dir $distance $maf_cutoff $normalization_method $independent_time_step_eqtl_dir $data_prep_version $num_pcs
fi





















##################################################################
### Run Combined Haplotype Test (CHT / WASP)
##################################################################










